sdk/nodejs/policy/server.ts

```diff
@@ -573,3 +586,102 @@ function isTypeOf<TResource extends Resource>(
+                if (enforcementLevel !== "remediate" || !isResourcePolicy(p) || !p.remediateResource) {
+                    continue;
+                }
+
+                const args: ResourceValidationArgs = {
+                    urn, name, type, opts, props,
+                    getConfig: makeGetConfigFun(p.name),
+                    isType: function <TResource extends Resource>(
+                        resourceClass: { new(...rest: any[]): TResource },
+                    ): boolean {
+                        return isTypeOf(type, resourceClass);
+                    },
+                    asType: function <TResource extends Resource, TArgs>(
+                        resourceClass: { new(name: string, args: TArgs, ...rest: any[]): TResource },
+                    ): Unwrap<NonNullable<TArgs>> | undefined {
+                        return isTypeOf(type, resourceClass)
+                            ? props as Unwrap<NonNullable<TArgs>>
+                            : undefined;
+                    },
+                };
+                const provider = getProviderResource(req);
+                if (provider) {
+                    args.provider = provider;
+                }
+
+                // Attempt to run the remediation; wrap this in a try block in case the user code throws.
+                let result: any = undefined;
+                let diagnostic: string | undefined = undefined;
+                try {
+                    // Pass the result of the validate call to Promise.resolve.
+                    // If the value is a promise, that promise is returned; otherwise
+                    // the returned promise will be fulfilled with the value.
+                    result = await Promise.resolve(p.remediateResource(args));
+                    if (result) {
+                        props = result;
+                    }
+                } catch (e) {
+                    const policyPack = `'${policyPackName}@${policyPackVersion}'`;
+                    const policyFrom = `remediation '${p.name}' from policy pack ${policyPack}`;
+                    if (e instanceof UnknownValueError) {
+                        diagnostic = `can't run ${policyFrom} during preview: ${e.message}`;
+                    } else {
+                        throw asGrpcError(e, `Error remediating resource with ${policyFrom}`);
+                    }
+                }
+
+                if (result || diagnostic) {
+                    // Remove the proxiness if it is there, to avoid unknowns causing issues.
+                    if (result && result.getOriginalTarget) {
+                        result = result.getOriginalTarget();
+                    }
+
+                    rs.push({
+                        policyName: p.name,
+                        policyPackName,
+                        policyPackVersion,
+                        description: p.description,
+                        properties: result,
```

Comment:
Don't we need to serialize the result? Turn assets/archives back into special objects, etc.?

Reply:
Indeed! I have added that as part of introducing secrets support.